plugins/node/opentelemetry-instrumentation-azure-functions/src/instrumentation.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { context as otelContext, propagation } from '@opentelemetry/api';
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationConfig,
+} from '@opentelemetry/instrumentation';
+import type * as azFunc from '@azure/functions';
+import { VERSION } from './version';
+
+export class AzureFunctionsInstrumentation extends InstrumentationBase {
+  constructor(config: InstrumentationConfig = {}) {
+    super('@opentelemetry/instrumentation-azure-functions', VERSION, config);
+  }
+
+  protected init() {
+    let disposable: azFunc.Disposable | undefined;
+
+    return new InstrumentationNodeModuleDefinition(
+      '@azure/functions',
+      ['^4.0.0'],
+      (moduleExports: typeof azFunc) => {
+        disposable = moduleExports.app.hook.preInvocation(context =>
+          this._preInvocationHook(context)
+        );
+        return moduleExports;
+      },
+      () => disposable?.dispose()
+    );
+  }
+
+  private _preInvocationHook(context: azFunc.PreInvocationContext): void {
+    const traceContext = context.invocationContext.traceContext;
+    if (traceContext) {
+      context.functionHandler = otelContext.bind(
+        propagation.extract(otelContext.active(), {
```

Comment:
I believe this only works if there is a W3CTraceContext propagator registered right?, this is not my area of expertise to be honest here, but I would expect that if you wanted to change current context you can do it using existing trace APIs, if you look at W3CTraceContextPropagator code, it will parse the traceParent and traceState, it will create a spanContext and then would be calling trace.setSpanContext

Reply:
I've never explicitly registered a W3CTraceContext propagator and it's worked just fine for me 🤷‍♂️. [Here's a sample app](https://github.com/ejizba/azure-functions-nodejs-opentelemetry/blob/advanced-previewpackage/src/index.ts) where I'm using this. I am using auto instrumentation stuff, so not sure if that pulls it in

It's a little weird because we don't want to _create_ a span context, just bind to an existing context. The Azure Functions Host (our shared .NET component) creates the span for every function invocation not us

Reply:
The W3CTraceContext is actually registered in BasicTracerProvider that NodeTracerProvider extends from, so not sure if there is an actual scenario where this is not available, I cannot find some actual documentation about how we expect people to work with the context, @pichlermarc do you know if such documentation exists?

Reply:
Looks similar to the wrapper we're using around http Azure Functions handlers to obtain e2e tracing: https://github.com/gunzip/tracing-test-app/blob/main/src/wrapper.ts#L72

Reply:
Correct me if I'm wrong, but my takeaway is that this is probably fine in the vast majority of cases. Thanks!